webapi/src/main/scala/org/knora/webapi/util/date/CalendarDateUtilV2.scala

```diff
@@ -0,0 +1,543 @@
+}
+
+object CalendarDateV2 {
+    /**
+      * Converts a Julian Day Number to a [[CalendarDateV2]].
+      *
+      * @param julianDay the Julian Day Number.
+      * @param precision the desired precision.
+      * @param calendarName the name of the calendar to be used.
+      * @return a [[CalendarDateV2]] with the specified precision and calendar.
+      */
+    def fromJulianDayNumber(julianDay: Int, precision: DatePrecisionV2, calendarName: CalendarNameV2): CalendarDateV2 = {
+        // Convert the Julian Day Number to a com.ibm.icu.util.Calendar.
+        val (calendar: Calendar, maybeEra: Option[DateEraV2]) = calendarName match {
+            // TODO: support calendars other than Gregorian and Julian.
+
+            case gregorianOrJulianName: CalendarNameGregorianOrJulian =>
+                val calendar: GregorianCalendar = new GregorianCalendar(TimeZone.GMT_ZONE, ULocale.ENGLISH)
+                calendar.setGregorianChange(CalendarDateUtilV2.getGregorianCalendarChangeDate(gregorianOrJulianName))
+                calendar.set(Calendar.JULIAN_DAY, julianDay)
+
+                val maybeGregorianEra = calendar.get(Calendar.ERA) match {
+                    case GregorianCalendar.AD => Some(DateEraCE)
+                    case GregorianCalendar.BC => Some(DateEraBCE)
+                    case _ => throw AssertionException("Unreachable code")
+                }
+
+                (calendar, maybeGregorianEra)
+        }
+
+        // Get the year, month, and day from the com.ibm.icu.util.Calendar.
+        // Note: in com.ibm.icu.util.Calendar, month is 0-based.
+
+        val year: Int = calendar.get(Calendar.YEAR)
+        val month: Int = calendar.get(Calendar.MONTH) + 1
+        val day: Int = calendar.get(Calendar.DAY_OF_MONTH)
+
+        // Return a CalendarDateV2 in the requested precision.
+        precision match {
+            case DatePrecisionYear =>
+                CalendarDateV2(
+                    calendarName = calendarName,
+                    year = year,
+                    maybeMonth = None,
+                    maybeDay = None,
+                    maybeEra = maybeEra
+                )
+
+            case DatePrecisionMonth =>
+                CalendarDateV2(
+                    calendarName = calendarName,
+                    year = year,
+                    maybeMonth = Some(month),
+                    maybeDay = None,
+                    maybeEra = maybeEra
+                )
+
+            case DatePrecisionDay =>
+                CalendarDateV2(
+                    calendarName = calendarName,
+                    year = year,
+                    maybeMonth = Some(month),
+                    maybeDay = Some(day),
+                    maybeEra = maybeEra
+                )
+        }
+    }
+
+    /**
+      * Parses a string representing a single date, without the calendar. This method does does not check that the
+      * date is valid in its calendar; to do that, call `toJulianDayRange` on it.
```

Comment:
The actual validity will only be checked if a `GregorianCalendar` is instantiated and leniency is set to false, right?

Reply:
Yes, and this happens in `toJulianDayRange`.